build-logic/cleanup/src/main/java/gradlebuild/cleanup/KillLeakingJavaProcesses.java

```diff
@@ -0,0 +1,248 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package gradlebuild.cleanup;
+
+import java.io.BufferedReader;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.PrintStream;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+/**
+ * NOTICE: this class is invoked via java command line, so we must NOT DEPEND ON ANY 3RD-PARTY LIBRARIES except JDK 11.
+ *
+ * Usage: java build-logic/cleanup/src/main/java/gradlebuild/cleanup/KillLeakingJavaProcesses.java
+ *
+ * Kill all Java processes matching:
+ * 1. Main class is `GradleWorkerMain`/`KotlinCompileDaemon`.
+ * 2. Main class is `GradleDaemon` and classpath is inside the current working directory.
+ */
+public class KillLeakingJavaProcesses {
+    private static final Pattern UNIX_JAVA_PROCESS_PATTERN = Pattern.compile("^(\\d+)\\s+(.*/bin/java)\\s+(.*)$");
+    private static final Pattern WINDOWS_JAVA_PROCESS_PATTERN = Pattern.compile("^\"?(.*[/\\\\]bin[/\\\\]java(\\.exe)?)\"\\s+(.*)\\s+(\\d+)$");
+    private static final Pattern MAIN_CLASS_PATTERN = Pattern.compile("([a-z]+\\.)+[A-Z]\\w+");
+
+    private static final String WORKING_DIRECTORY_PATH = new File("").getAbsolutePath();
+
+    public static void main(String[] args) {
+        System.out.println("Working dir is: " + WORKING_DIRECTORY_PATH);
+        ps().stream()
+            .peek(javaProcessInfo -> System.out.println("Found java process " + javaProcessInfo))
+            .filter(JavaProcessInfo::isLeakingJavaProcess)
+            .forEach(JavaProcessInfo::kill);
+
+        System.out.println("After cleanup:");
+        System.out.println(run(determinePsCommand()));
+    }
+
+    private static List<JavaProcessInfo> ps() {
+        ExecResult result = run(determinePsCommand()).assertZeroExit();
```

Comment:
Can we use the `ProcessHandle` api here, so we avoid calling any command line commands altogether?